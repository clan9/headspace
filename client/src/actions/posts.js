import axios from 'axios';
import {
  FETCH_ALL_POSTS,
  FETCH_POST,
  ADD_POST,
  EDIT_POST,
  DELETE_POST,
  ADD_COMMENT,
  LIKE_POST,
  UNLIKE_POST,
  POST_ERROR
} from 'actions/types';

export const fetchAllPosts = () => async dispatch => {
  try {
    const res = await axios.get('/api/posts');
    dispatch({ type: FETCH_ALL_POSTS, payload: res.data });
  } catch (err) {
    dispatch({
      type: POST_ERROR,
      payload: err.response.data.msg
    });
  }
};

export const fetchPost = id => async dispatch => {
  try {
    const res = await axios.get(`/api/posts/${id}`);

    dispatch({ type: FETCH_POST, payload: res.data });
  } catch (err) {
    console.log('ERROR MSG:', err);
    dispatch({
      type: POST_ERROR,
      payload: err.response.data.msg
    });
  }
};

export const addPost = (formData, history) => async dispatch => {
  try {
    const res = await axios.post('/api/posts', formData);
    dispatch({ type: ADD_POST, payload: res.data });
    history.push('/dashboard');
  } catch (err) {
    console.log('ERROR MSG:', err.response.data.msg);
    dispatch({
      type: POST_ERROR,
      payload: err.response.data.msg
    });
  }
};

export const deletePost = (id, history) => async dispatch => {
  try {
    const res = await axios.delete(`/api/posts/${id}`);
    dispatch({ type: DELETE_POST, payload: res.data });
    history.push('/dashboard');
  } catch (err) {
    console.log('ERROR MSG:', err.response.data.msg);
    dispatch({
      type: POST_ERROR,
      payload: err.response.data.msg
    });
  }
};

export const editPost = (id, formData, history) => async dispatch => {
  try {
    const res = await axios.patch(`/api/posts/${id}`, formData);
    dispatch({ type: EDIT_POST, payload: res.data });
    history.push('/dashboard');
  } catch (err) {
    console.log('ERROR MSG:', err.response.data.msg);
    dispatch({
      type: POST_ERROR,
      payload: err.response.data.msg
    });
  }
};

export const addComment = ({
  text,
  createdAt,
  id,
  history
}) => async dispatch => {
  try {
    const res = await axios.post(`/api/posts/${id}/comments`, {
      text,
      createdAt
    });
    // res.data is the post, not just the comments array for that post (this is a change to original controller)
    dispatch({ type: ADD_COMMENT, payload: res.data });
    history.push(`/post/${id}`);
  } catch (err) {
    console.log('ERROR MSG:', err.response.data.msg);
    dispatch({
      type: POST_ERROR,
      payload: err.response.data.msg
    });
  }
};

// Include this or not?
export const deleteComment = () => async dispatch => {
  return {};
};

export const likePost = id => async dispatch => {
  try {
    const res = await axios.post(`/api/posts/${id}/like`);
    dispatch({ type: LIKE_POST, payload: res.data });
  } catch (err) {
    dispatch({
      type: POST_ERROR,
      payload: err.response.data.msg
    });
    setTimeout(() => {
      dispatch({ type: POST_ERROR, payload: '' });
    }, 6000);
  }
};

export const unlikePost = id => async dispatch => {
  try {
    const res = await axios.delete(`/api/posts/${id}/unlike`);
    dispatch({ type: UNLIKE_POST, payload: res.data });
  } catch (err) {
    dispatch({
      type: POST_ERROR,
      payload: err.response.data.msg
    });
    setTimeout(() => {
      dispatch({ type: POST_ERROR, payload: '' });
    }, 6000);
  }
};
