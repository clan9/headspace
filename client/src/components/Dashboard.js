import React, { useEffect } from 'react';
import PostList from 'components/PostList';
import PostFilter from 'components/PostFilter';
import { connect } from 'react-redux';

export const Dashboard = ({ name = '' }) => {
  const nameCasing = () => {
    if (name) {
      return name.charAt(0).toUpperCase() + name.slice(1);
    }
  };

  useEffect(() => {
    window.scrollTo(0,0)
  }, [])

  return (
    <div className="content-container">
      <h2 className="page__heading">Welcome {nameCasing()}</h2>
      {!name && (
        <p className="dashboard__text">
          In order to submit a post, comment or like, please sign up for a new
          account or log in.
        </p>
      )}
      <PostFilter />
      <PostList />
    </div>
  );
};

const mapStatetoProps = state => ({
  name: state.profile.name
});

export default connect(mapStatetoProps)(Dashboard);
