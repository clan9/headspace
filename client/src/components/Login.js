import React from 'react';
import { Link } from 'react-router-dom';

const Login = props => {
  return (
    <div className="login">
      <div className="login__box">
        <h1 className="login__title">headSpace</h1>
        <p className="login__text">
          Share your thoughts and connect with others!
        </p>
        <Link to="/signup" className="login__buttons cta">
          Sign Up!
        </Link>
        <Link to="/signin" className="login__buttons">
          Log In
        </Link>
        <Link to="/dashboard" className="login__buttons">
          Browse Posts
        </Link>
      </div>
    </div>
  );
};

export default Login;
