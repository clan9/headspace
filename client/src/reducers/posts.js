import {
  FETCH_ALL_POSTS,
  FETCH_POST,
  ADD_POST,
  EDIT_POST,
  DELETE_POST,
  ADD_COMMENT,
  POST_ERROR,
  LIKE_POST,
  UNLIKE_POST
} from 'actions/types';

const initialState = {
  posts: [],
  post: {},
  loading: true,
  error: ''
};

export default (state = initialState, action) => {
  const { type, payload } = action;

  switch (type) {
    case FETCH_ALL_POSTS:
      return { ...state, posts: payload, post: {}, loading: false };
    case FETCH_POST:
      return { ...state, post: payload, loading: false };
    case ADD_POST:
      return {
        ...state,
        posts: [...state.posts, payload],
        loading: false
      };
    case DELETE_POST:
      return {
        ...state,
        posts: state.posts.filter(post => post._id !== payload._id),
        post: {},
        loading: false
      };
    case EDIT_POST:
      return {
        ...state,
        posts: state.posts.map(post => {
          if (post._id === payload._id) {
            return { ...post, ...payload };
          } else {
            return post;
          }
        }),
        post: {},
        loading: false
      };
    case ADD_COMMENT:
    case LIKE_POST:
    case UNLIKE_POST:
      return {
        ...state,
        posts: state.posts.map(post => {
          return post._id === payload._id ? { ...post, ...payload } : post;
        }),
        loading: false,
        error: ''
      };
    case POST_ERROR:
      return {
        ...state,
        error: payload,
        loading: false
      };
    default:
      return state;
  }
};
