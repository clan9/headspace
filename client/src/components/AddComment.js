import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter, Link } from 'react-router-dom';
import moment from 'moment';
import { addComment } from 'actions/posts';

export class AddComment extends Component {
  state = { text: '', createdAt: moment(), error: '' };

  onChange = e => {
    const text = e.target.value;
    this.setState(() => ({ text }));
  };

  resetError = () => {
    setTimeout(() => {
      this.setState(() => ({ error: '' }));
    }, 6000);
  };

  componentWillUnmount() {
    this.setState(() => ({ error: '' }))
  }

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.text) {
      this.setState(() => ({
        error: 'You cannot add an empty comment'
      }));
      this.resetError();
    } else {
      this.setState(() => ({ error: '' }));
      this.props.addComment({
        text: this.state.text,
        createdAt: this.state.createdAt.valueOf(),
        id: this.props.id,
        history: this.props.history
      });
    }
  };

  render() {
    return (
      <div className="content-container">
        <h2 className="page__heading">Add a Comment</h2>
        <form onSubmit={this.onSubmit} className="form">
          <div className="form__error__container">
            {this.state.error && (
              <p className="form__error">{this.state.error}</p>
            )}
          </div>

          <textarea
            type="text"
            placeholder="What did you think?"
            className="form__textarea"
            value={this.state.text}
            onChange={this.onChange}
          />
          <div className="comment__form__button__container">
            <Link to="/dashboard" className="comment__form__button">
              Home
            </Link>
            <button className="comment__form__button" type="submit">
              Save Comment
            </button>
          </div>
        </form>
      </div>
    );
  }
}

AddComment.propTypes = {
  addComment: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) => ({
  post: state.posts.post,
  id: ownProps.match.params.id
});

export default connect(
  mapStateToProps,
  { addComment }
)(withRouter(AddComment));
