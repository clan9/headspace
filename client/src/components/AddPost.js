import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { addPost } from 'actions/posts';
import PostForm from 'components/PostForm';

export const AddPost = ({ addPost, history }) => {
  return (
    <div className="content-container">
      <h2 className="page__heading">Add a Post</h2>
      <PostForm
        onSubmit={post => {
          addPost(post, history);
        }}
      />
    </div>
  );
};

AddPost.propTypes = { addPost: PropTypes.func.isRequired };

export default connect(
  null,
  { addPost }
)(withRouter(AddPost));
