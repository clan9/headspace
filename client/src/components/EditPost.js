import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { editPost } from 'actions/posts';
import PostForm from 'components/PostForm';

export const EditPost = ({ post, id, history, editPost }) => {
  return (
    <div className="content-container">
      <h2 className="page__heading">Edit your Post</h2>
      <PostForm
        post={post}
        onSubmit={post => {
          editPost(id, post, history);
        }}
      />
    </div>
  );
};

EditPost.propTypes = {
  editPost: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) => ({
  post: state.posts.post,
  id: ownProps.match.params.id
});

export default connect(
  mapStateToProps,
  { editPost }
)(withRouter(EditPost));
