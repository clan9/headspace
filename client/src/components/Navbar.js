import React from "react";
import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";
import { connect } from "react-redux";
import { logout } from "actions/auth";

export const Navbar = ({ logout, auth: { isAuthenticated, loading } }) => {
  const userLogout = () => {
    logout();
  };

  const renderLinks = () => {
    if (!isAuthenticated && !loading) {
      return (
        <div className="navbar">
          <div className="navbar__left">
            <NavLink
              to="/dashboard"
              activeClassName="is-active"
              className="navbar__link navbar__link--logo"
            >
              headSpace
            </NavLink>
          </div>

          <div className="navbar__right">
            <NavLink
              to="/signup"
              activeClassName="is-active"
              className="navbar__link"
            >
              Sign Up
            </NavLink>
            <NavLink
              to="/signin"
              activeClassName="is-active"
              className="navbar__link"
            >
              Log In
            </NavLink>
          </div>
        </div>
      );
    } else {
      return (
        <div className="navbar">
          <div className="navbar__left">
            <NavLink
              to="/dashboard"
              activeClassName="is-active"
              className="navbar__link navbar__link--logo"
            >
              headSpace
            </NavLink>
          </div>
          <div className="navbar__right">
            <NavLink
              to="/profile"
              activeClassName="is-active"
              className="navbar__link"
            >
              Your Stats
            </NavLink>
            <NavLink to="/" onClick={userLogout} className="navbar__link">
              Logout
            </NavLink>
          </div>
        </div>
      );
    }
  };

  return <div>{renderLinks()}</div>;
};

Navbar.propTypes = {
  logout: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { logout }
)(Navbar);
