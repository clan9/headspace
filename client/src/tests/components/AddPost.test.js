import React from 'react';
import { shallow } from 'enzyme';
import { AddPost } from 'components/AddPost';
import posts from 'tests/fixtures/posts';

describe('AddPost component', () => {
  let addPost, history, wrapper;

  beforeEach(() => {
    addPost = jest.fn();
    history = jest.fn();
    wrapper = shallow(<AddPost addPost={addPost} history={history} />);
  });

  it('should render the component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle onSubmit', () => {
    console.log(wrapper.debug());
    // wrapper.find('PostForm').prop('onSubmit')(posts[0]);
    // expect(addPost).toHaveBeenCalled();
  });
});
