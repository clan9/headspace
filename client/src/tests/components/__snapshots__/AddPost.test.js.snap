// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPost component should render the component correctly 1`] = `
<div
  className="content-container"
>
  <h2
    className="page__heading"
  >
    Add a Post
  </h2>
  <Connect(PostForm)
    onSubmit={[Function]}
  />
</div>
`;
